crates/typst-cli/src/query.rs

```diff
@@ -0,0 +1,133 @@
+
+    let mut tracer = Tracer::default();
+    let result = typst::compile(&world, &mut tracer);
+    let duration = start.elapsed();
+    let warnings = tracer.warnings();
+
+    match result {
+        // Print metadata
+        Ok(document) => {
+            let introspector = Introspector::new(&document.pages);
+
+            if let Some(key) = &command.key {
+                let provided_metadata = introspector
+                    .query(&keyvalue_selector(key))
+                    .iter()
+                    .filter_map(|c| c.field("value"))
+                    .collect::<Vec<_>>();
+                format(&provided_metadata, &command)?;
+            }
+
+            if let Some(selector) = &command.selector {
+                let selected_metadata = introspector
+                    .query(&generic_selector(selector, &world)?)
+                    .into_iter()
+                    .map(|x| SelectedElement {
+                        typename: x.func().name().into(),
+                        attributes: x
+                            .clone()
+                            .into_inner()
+                            .fields()
+                            .map(|(k, v)| (k.clone(), v))
+                            .collect(),
+                    })
+                    .collect::<Vec<_>>();
+                format(&selected_metadata, &command)?;
+            }
+
+            tracing::info!("Processing succeeded in {duration:?}");
+
+            print_diagnostics(&world, &[], &warnings, command.common.diagnostic_format)
+                .map_err(|_| "failed to print diagnostics")?;
+        }
+
+        // Print diagnostics.
+        Err(errors) => {
+            set_failed();
+            tracing::info!("Processing failed");
+
+            print_diagnostics(
+                &world,
+                &errors,
+                &warnings,
+                command.common.diagnostic_format,
+            )
+            .map_err(|_| "failed to print diagnostics")?;
+        }
+    }
+
+    Ok(())
+}
+
+fn generic_selector(description: &str, world: &dyn World) -> StrResult<Selector> {
+    let evaluated_selector = eval_string(
+        world.track(),
+        &format!("selector({description})"),
+        Span::detached(),
+        EvalMode::Code,
+        Scope::default(),
+    )
+    .map_err(|_| "Error evaluating the selector string.")?;
+
+    let Dyn(selector) = evaluated_selector else {
+        bail!("Parsing of selector string not successfully.")
+    };
+
+    Ok(selector.downcast::<Selector>().unwrap().to_owned())
+}
+
+fn keyvalue_selector(key: &str) -> Selector {
+    let mut bounds = Dict::new();
+    bounds.insert("key".into(), key.into_value());
+
+    Selector::Elem(ProvideElem::func(), Some(bounds))
+}
+
+fn format<T: Serialize>(data: &[T], command: &QueryCommand) -> StrResult<()> {
+    if command.one && data.len() != 1 {
+        bail!("One piece of metadata expected, but {} found.", data.len())
```

Comment:
error messages should be all lowercase and without a ending period (it will be added by the error handler)